src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CustomAttributeBasedDependencyAlgorithm.cs

```diff
@@ -252,11 +252,17 @@ private static bool AddDependenciesFromCustomAttributeArgument(DependencyList de
             if (factory.MetadataManager.IsReflectionBlocked(typeofType))
                 return false;
 
-            // We go for the entire EEType because the reflection stack at runtime might need an EEType anyway
-            // (e.g. if this is a `typeof(SomeType[,,])` just having the metadata for SomeType is not enough).
-            // Theoretically, we only need the EEType if this is a constructed type, but let's go for the full
-            // EEType for the sake of consistency.
-            dependencies.Add(factory.MaximallyConstructableType(typeofType), "Custom attribute blob");
+            // Grab the metadata nodes that will be necessary to represent the typeof in the metadata blob
+            TypeMetadataNode.GetMetadataDependencies(ref dependencies, factory, typeofType, "Custom attribute blob");
+
+            if (typeofType.IsArray
+                || (typeofType.HasInstantiation && !typeofType.IsGenericDefinition))
+            {
+                // We go for the entire EEType because the reflection stack at runtime will need an EEType anyway
```

Comment:
Is there a good place in the reflection stack that shows the dependency on EEType and that can be linked from here?

It looks odd to me that the reflection stack does not need EETypes for regular types, but it does need them for arrays and generics.

Reply:
IIRC is this one for arrays:

https://github.com/dotnet/runtimelab/blob/2c625048711145c22780dffd8f6a6635aa245c4b/src/coreclr/nativeaot/System.Private.Reflection.Core/src/System/Reflection/Runtime/General/TypeUnifier.cs#L278-L289

And this one for generic types:

https://github.com/dotnet/runtimelab/blob/2c625048711145c22780dffd8f6a6635aa245c4b/src/coreclr/nativeaot/System.Private.Reflection.Core/src/System/Reflection/Runtime/General/TypeUnifier.cs#L426-L443

(The `// We only permit` part)

If we just go and delete those lines things will probably still work.

The intent of that code was to blame `MakeGenericType`/`MakeArrayType` for reflection failures at runtime instead of having these APIs succeed and getting an exception once we actually want to do something with the type (call static method, activate, etc.).

I think this behavior makes sense for user code because this API is the problem.

It's annoying for framework code because e.g. without the special casing in the compiler we will get an exception when trying to `GetCustomAttributeData` for the attribute.

Reply:
On a second thought, maybe the compiler doesn't need to do this special casing here. This situation is the same as what we would get when e.g. querying fields:

```csharp
using System;

ClassWithField.Klass = new ClassWithField() { MyGen = null };

typeof(Gen<>).ToString();

foreach (var f in typeof(ClassWithField).GetFields(System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public))
{
    Console.WriteLine(f.FieldType.FullName);
}

class ClassWithField
{
    public static ClassWithField Klass;

    public Gen<int> MyGen;
}

class Gen<T> { }
```

This will fail with:

```
Unhandled Exception: System.Reflection.MissingMetadataException: 'Gen<System.Int32>' is missing metadata. For more information, please visit http://go.microsoft.com/fwlink/?LinkID=392859
   at System.Reflection.Runtime.TypeInfos.RuntimeConstructedGenericTypeInfo.ConstructedGenericTypeTable.Factory(RuntimeConstructedGenericTypeInfo.UnificationKey) + 0x24b
   at System.Collections.Concurrent.ConcurrentUnifierWKeyed`2.GetOrAdd(K) + 0x23f
   at System.Reflection.Runtime.TypeInfos.RuntimeConstructedGenericTypeInfo.GetRuntimeConstructedGenericTypeInfo(RuntimeTypeInfo, RuntimeTypeInfo[], RuntimeTypeHandle) + 0x6f
   at System.Reflection.Runtime.TypeInfos.RuntimeConstructedGenericTypeInfo.GetRuntimeConstructedGenericTypeInfo(RuntimeTypeInfo, RuntimeTypeInfo[]) + 0x56
   at System.Reflection.Runtime.General.TypeUnifier.GetConstructedGenericType(RuntimeTypeInfo, RuntimeTypeInfo[]) + 0x2a
   at System.Reflection.Runtime.General.TypeResolver.TryResolveTypeSignature(TypeSpecificationHandle, MetadataReader, TypeContext, Exception&) + 0x945
   at System.Reflection.Runtime.General.TypeResolver.TryResolve(Handle, MetadataReader, TypeContext, Exception&) + 0x160
   at System.Reflection.Runtime.General.TypeResolver.Resolve(Handle, MetadataReader, TypeContext) + 0x61
   at System.Reflection.Runtime.FieldInfos.NativeFormat.NativeFormatRuntimeFieldInfo.get_FieldRuntimeType() + 0x93
   at System.Reflection.Runtime.FieldInfos.RuntimeFieldInfo.get_FieldType() + 0x70
   at <Program>$.<Main>$(String[]) + 0x130
```

The compiler doesn't try to go out of its way to make sure we can access the field type either. Maybe that's fine?

Reply:
I do not have opinion, except that consistency and less special cases is good.

Reply:
Okay, let's see if we can get away without it. We can always restore that behavior.